File: ising.jl
using Printf
using LinearAlgebra


abstract type Parameters end

struct params <: Parameters
    L::Integer          # system size
    J::Float            # coupling
    kT::Float           # temperature
    thermal::Integer    # thermalize
    Nmc::Integer        # no. of sweeps
end

function init!(params::Parameters)
    state = zeros(params.L, params.L)
    for i=1:params.L, j=1:params.L
            state[i][j] = (rand() > 0.5 ? 1 : -1)
    end
    state
end

function neighbors(state, i, j)
    L = size(state)[1]
    left = (i+1)%L == 0 ? state[L][j] : state[i+1][j]
    right = (i-1)%L == 0 ? state[1][j] : state[i-1][j]
    up = (j+1)%L == 0 ? state[i][1] : state[i][j+1]
    down = (j-1)%L == 0 ? state[i][L] : state[i][j-1]
    return left+right+up+down
end

function energy(params::Parameters, state::Matrix)
	E::Float64 = 0
    L = size(state)[1]
    for i=1:L, j=1:L
        E -= params.J*state[i][j]*neighbors(state, i, j)
    end
    E
end


function Magnetization(model)
    m::Float64 = 0
    N = length(model.state)
    for i in 1:N
        m += model.state[i]
    end
    return m
end


function thermalize!(model, kT)
	N = length(model.state)
	for i=1:model.parameters.thermalize
		for j=1:N
			site = convert(Int64, 1+round(rand()*(N-1), digits=0))
			ΔE = 2*model.parameters.J1*model.state[site]*sumNeighbors(model, site)
			if ΔE < 0 || rand() < exp(-ΔE/kT)
				model.state[site] *= -1
			end
		end
	end
	return model
end




function run!(model)
	N = length(model.state)
	observables = Observables(model.parameters.temps)
	Temps = LinRange(model.parameters.tmax, model.parameters.tmin, model.parameters.temps)
	totalTime = time_ns()
	for (t, T) in enumerate(Temps)
		@printf "\n %d of %d Temps | Starting MC at T = %0.5f eV\n" t length(Temps) T
		@printf "Starting thermalization...\n"
		warmTime = time_ns()
		model = thermalize!(model, T)
		@printf "Finished warm up in %.4f s\n" (time_ns() - warmTime)/1.0e9
		@printf " T        E        |Mz|        Cv        χ\n"
		etot = e2tot = mtot = m2tot = mabstot = 0.0
		for n=1:model.parameters.mccycles
			model.energy = Energy(model)
			model.magnet = Magnetization(model)
			for s=1:N
				site = convert(Int64, 1+round(rand()*(N-1), digits = 0))
				ΔE   = 2*model.parameters.J1*model.state[site]*sumNeighbors(model, site)
				if ΔE < 0 || rand() < exp(-ΔE/T)
					model.state[site] *= -1
					model.energy += ΔE
					model.magnet += 2*model.state[site]
				end
			end
			etot += model.energy/2.0
			e2tot += (model.energy/2.0)^2
			mtot += model.magnet
			m2tot += model.magnet * model.magnet
			mabstot += sqrt(model.magnet * model.magnet)
		end
        	E_avg    = etot/convert(Float64, model.parameters.mccycles*N)
        	E2_avg   = e2tot/convert(Float64, model.parameters.mccycles*N)
        	M_avg    = mtot/convert(Float64, model.parameters.mccycles*N)
        	M2_avg   = m2tot/convert(Float64, model.parameters.mccycles*N)
        	Mabs_avg = mabstot/convert(Float64, model.parameters.mccycles*N)

		observables.temperature[t] = T
		observables.energy[t] = E_avg		
		observables.magnetization[t] = Mabs_avg 		
        	observables.heatcapacity[t] = (E2_avg - E_avg*E_avg*N)/convert(Float64, T*T*model.parameters.mccycles*N)
        	observables.susceptibility[t] = (M2_avg - M_avg*M_avg*N)/convert(Float64, T*model.parameters.mccycles*N)
		@printf "\n %0.4f        %0.4f        %0.4f        %0.4f        %0.4f" T E_avg Mabs_avg observables.heatcapacity[t] observables.susceptibility[t] 	
	end
	@printf "\n Finished MC for all temperatures requested in %0.4f seconds" (time_ns() - totalTime)/1.0e9
	return model, observables
end

File: ising.py
#!/usr/bin/env python
"""
Two-dimensional Ising model on a square lattice in python.
"""
import sys
import numpy as np
from datetime import datetime


class ising:
    def __init__(self, L, kT, J, thermal, Nmc):
        self.L = L
        self.kT = kT
        self.J = J
        self.thermal = thermal
        self.Nmc = Nmc

    def init(self):
        state = np.zeros((self.L, self.L))
        for i in range(self.L):
            for j in range(self.L):
                if np.random.rand() > 0.5:
                    state[i][j] = 1
                else:
                    state[i][j] = -1
        return state

    def neighbors(self, i, j):
        left = self.state[(i-1)%self.L][j]
        right = self.state[(i+1)%self.L][j]
        up = self.state[i][(j-1)%self.L]
        down = self.state[i][(j+1)%self.L]
        return left+right+up+down

    def magnetization(self):
        return np.mean(self.state)

    def energy(self):
        E = 0
        for i in range(len(self.state)):
            for j in range(len(self.state)):
                E -= J*self.state[i][j]*self.neighbors(i, j)
        return E

    def mc(self):
        self.state = self.init() 
        Eavg = E2avg = M = Mavg = 0
        for n in range(self.thermal + self.Nmc):
            E = self.energy()
            M = self.magnetization()
            for s in range(self.L**2):
                row = int(np.random.random()*self.L)
                col = int(np.random.random()*self.L)
                spin = self.state[row, col]
                dE = 2*J*spin*self.neighbors(row, col)
                prob = np.exp(-dE/self.kT)
                if dE <= 0 or np.random.random() <= prob:
                    self.state[row, col] = -1*self.state[row, col]
                    if n >= self.thermal:
                        E += dE
                        M += 2*self.state[row, col]

            if n >= self.thermal:
                Eavg += E
                E2avg += E**2
                Mavg += M

        Eavg /= float(self.Nmc)
        E2avg /= float(self.Nmc)

        Evar = (E2avg - Eavg*E_avg)/float(self.L*self.L*sefl.kT*self.kT)
        Mavg /= float(self.Nmc*self.L*self.L)
        Eavg /= float(self.L*self.L)
        return Eavg, Evar, Mavg


if __name__ == "__main__":
    # default params
    L = 10
    kT = 4.0
    J = 1.0
    thermal = 10000
    Nmc = 100000
    help = """2d ising model on a square lattice. Command line arguments are:
            -------------------------------------------
             Variable         |   Default  |     Tag 
             -------------------------------------------
             L (system size)  |   10       |     -L
             kT (temperature) |   4.0      |     -kT
             J  (coupling)    |   1.0      |     -J
             thermal (warmup) |   10000    |     -thermal
             Nmc (no. sweeps) |  100000    |     -Nmc
             -------------------------------------------"""
    for arg in range(1, len(sys.argv)):
        if sys.argv[arg] == "-L":
            L = sys.argv[arg+1]
        elif sys.argv[arg] == "-kT":
            kT = sys.argv[arg+1]
        elif sys.argv[arg] == "-J":
            J = sys.argv[arg+1]
        elif sys.argv[arg] == "-thermal":
            thermal = sys.argv[arg+1]
        elif sys.argv[arg] == "-Nmc":
            Nmc = sys.argv[arg+1]
        elif sys.argv[arg] == "-h":
            print(help)
            sys.exit(0)
    model = ising(L, kT, J, thermal, Nmc)
    print("running 2d ising model at {}...".format(str(datetime.now())))
    model.mc()
